code/javascript/js-assignment-7-classes.js

```diff
@@ -0,0 +1,34 @@
+class BankAccount {
+  constructor(ownerName) {
+    this.ownerName = ownerName;
+    this.balance = 0;
+  }
+
+//const person = new BankAccount("Ngozi Clara");
+//console.log(person);
+
+// Challenge 2: Create class methods
+
+  deposit(amount) {
+    this.balance += amount;
+  }
+
+  withdraw(amount) {
+    if (this.balance >= amount) {
+      this.balance -= amount;
+    } else {
+      console.log("Insufficient Funds");
+    }
+  }
+}
+
+const person = new BankAccount("Ngozi Clara");
+
+person.deposit(100);
+console.log(person.balance);
+
+person.withdraw(50);
+console.log(person.balance);
+
+person.withdraw(60);
+console.log(person.balance);
```

Comment:
The provided code defines a `BankAccount` class with methods for depositing and withdrawing money. Here are some improvements and enhancements that can be made:

1. **Initial Balance**: Allow setting an initial balance when creating the account.
2. **Transaction Logging**: Log transactions for better tracking.
3. **Validation**: Add validation for negative amounts in deposit and withdrawal methods.
4. **Error Handling**: Use more specific error messages.
5. **Balance Inquiry**: Add a method to display the current balance.
6. **ToString Method**: Provide a `toString` method to give a summary of the account.

Here's the improved version:

```javascript
class BankAccount {
  constructor(ownerName, initialBalance = 0) {
    this.ownerName = ownerName;
    this.balance = initialBalance;
    this.transactions = []; // To keep track of transactions
  }

  // Method to deposit money
  deposit(amount) {
    if (amount <= 0) {
      console.log("Deposit amount must be positive");
      return;
    }
    this.balance += amount;
    this.transactions.push({ type: "Deposit", amount });
    console.log(`Deposited ${amount}. New balance: ${this.balance}`);
  }

  // Method to withdraw money
  withdraw(amount) {
    if (amount <= 0) {
      console.log("Withdrawal amount must be positive");
      return;
    }
    if (this.balance >= amount) {
      this.balance -= amount;
      this.transactions.push({ type: "Withdrawal", amount });
      console.log(`Withdrew ${amount}. New balance: ${this.balance}`);
    } else {
      console.log("Insufficient Funds");
    }
  }

  // Method to check the balance
  getBalance() {
    console.log(`Current balance: ${this.balance}`);
    return this.balance;
  }

  // Method to print account summary
  toString() {
    return `Owner: ${this.ownerName}, Balance: ${this.balance}`;
  }

  // Method to print all transactions
  printTransactions() {
    console.log("Transactions:");
    this.transactions.forEach((transaction, index) => {
      console.log(`${index + 1}. ${transaction.type} - ${transaction.amount}`);
    });
  }
}

// Creating an instance of BankAccount
const person = new BankAccount("Ngozi Clara", 50);

// Deposit money
person.deposit(100);
person.getBalance();

// Withdraw money
person.withdraw(50);
person.getBalance();

// Attempt to withdraw more than the balance
person.withdraw(150);
person.getBalance();

// Attempt to deposit a negative amount
person.deposit(-10);

// Print account summary
console.log(person.toString());

// Print all transactions
person.printTransactions();
```

### Key Improvements:
1. **Initial Balance**: The constructor accepts an initial balance, defaulting to 0 if not provided.
2. **Transaction Logging**: A `transactions` array is added to log each deposit and withdrawal, providing a history of transactions.
3. **Validation**: The `deposit` and `withdraw` methods include checks for positive amounts, ensuring only valid transactions are processed.
4. **Error Handling**: More descriptive messages are provided for invalid actions, like trying to withdraw more than the balance or depositing a negative amount.
5. **Balance Inquiry**: The `getBalance` method allows checking the current balance.
6. **Account Summary**: The `toString` method provides a quick summary of the account, including the owner and balance.
7. **Transaction History**: The `printTransactions` method outputs a list of all transactions, offering transparency and accountability.

This code structure makes the `BankAccount` class more robust and user-friendly.